### PR TITLE
Button Menus (Code Review)

### PR DESCRIPTION
**Pull request by zcmoore**

### CODE REVIEW

File: src/edu/asu/plp/tool/prototype/Main.java [hunk 1]
```diff
@@ -413,45 +463,843 @@ private ProjectExplorerTree createProjectTree()
...
 		toolbar.setSpacing(5);
 		ObservableList<Node> buttons = toolbar.getChildren();
 		
 		EventHandler<MouseEvent> listener;
 		Node button;
 		
+		DropShadow lBlueShadow = new DropShadow();
+		lBlueShadow.setColor(Color.LIGHTBLUE);
+		DropShadow dBlueShadow = new DropShadow();
+		dBlueShadow.setColor(Color.DARKBLUE);
```
Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/Main.java]: Style: prefer to spell-out "light" and "dark" but this isn't a big issue

File: src/edu/asu/plp/tool/prototype/Main.java [hunk 2]
```diff
@@ -413,45 +463,843 @@ private ProjectExplorerTree createProjectTree()
...
 		// TODO: replace event handlers with actual content
+		Node projectButton = new ImageView("toolbar_new.png");
+		projectButton.addEventHandler(MouseEvent.MOUSE_ENTERED,
+				new EventHandler<MouseEvent>() {
+					@Override
+					public void handle(MouseEvent e)
+					{
+						projectButton.setEffect(lBlueShadow);
+					}
+				});
```
Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/Main.java]: This would be cleaner as a functional argument. Example:
```
projectButton.addEventHandler((event) -> projectButton.setEffect(lBlueShadow));
```
or
```
projectButton.addEventHandler(this::onNewProjectClicked);
```
with the following method:
```
public void onNewProjectClicked(MouseEvent event)
{
    projectButton.setEffect(lBlueShadow);
}
```

File: src/edu/asu/plp/tool/prototype/Main.java [hunk 3]
```diff
@@ -413,45 +463,843 @@ private ProjectExplorerTree createProjectTree()
...
+		projectButton.addEventHandler(MouseEvent.MOUSE_RELEASED,
+				new EventHandler<MouseEvent>() {
+					@Override
+					public void handle(MouseEvent e)
+					{
+						projectButton.setEffect(lBlueShadow);
+					}
+				});
```
Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/Main.java]: These would also be cleaner as a functional arguments. See above.

File: src/edu/asu/plp/tool/prototype/Main.java [hunk 4]
```diff
@@ -413,45 +463,843 @@ private ProjectExplorerTree createProjectTree()
...
 		listener = this::onOpenProjectClicked;
 		button.setOnMouseClicked(listener);
 		buttons.add(button);
 		
+		buttons.add(new Separator(Orientation.VERTICAL));
+		
 		button = new ImageView("toolbar_save.png");
-		listener = (event) -> console.println("Save Project Clicked");
+		listener = this::onSaveProjectClicked;
 		button.setOnMouseClicked(listener);
```
Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/Main.java]: This looks like something I did, but it might look better on one line as follows:
```
button.setOnMouseClicked(this::onSaveProjectClicked);
```

File: src/edu/asu/plp/tool/prototype/Main.java [hunk 5]
```diff
@@ -413,45 +463,843 @@ private ProjectExplorerTree createProjectTree()
...
+		listener = (event) -> console.println("GPIO Clicked");
+		button.setOnMouseClicked(listener);
+		buttons.add(button);
+		
+		button = new ImageView("toolbar_exclamation.png");
+		listener = (event) -> console.println("Interupt Clicked");
+		button.setOnMouseClicked(listener);
+		buttons.add(button);
+		
+		for (int x = 9; x <= 23; x++)
```
Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/Main.java]: These numbers constitute black-magic. According to the Righteous Magicians Act of 1999, we can't allow dark magic to exist...

Replace them with constants OR (better yet) add the buttons that need to be enabled/disabled to a Set, and iterate over that set.

Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/Main.java]: In fact, just do the set. constants in this case are really dirty and prone to errors in the future.

File: src/edu/asu/plp/tool/prototype/Main.java [hunk 6]
```diff
@@ -413,45 +463,843 @@ private ProjectExplorerTree createProjectTree()
...
+		itemNew.setOnAction((event) -> {
+			createNewProject();
+		});
+		MenuItem itemOpen = new MenuItem("Open PLP Project");
+		itemOpen.setGraphic(new ImageView(new Image("toolbar_open.png")));
+		itemOpen.setAccelerator(
+				new KeyCodeCombination(KeyCode.O, KeyCombination.CONTROL_DOWN));
+		itemOpen.setOnAction((event) -> {
+			// TODO: Add Event for menu item
+		});
```
Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/Main.java]: A single blank line between these blocks would help readability
(between lines 655&666, 662&663, etc)

File: src/edu/asu/plp/tool/prototype/Main.java [hunk 7]
```diff
@@ -413,45 +463,843 @@ private ProjectExplorerTree createProjectTree()
...
+		help.getItems().addAll(itemQuickRef, itemOnlineManual, new SeparatorMenuItem(),
+				itemReportIssue, itemGoogleIssues, new SeparatorMenuItem(), itemAboutPLP,
+				itemSWLicense);
+				
+		menuBar.getMenus().addAll(file, edit, view, project, tools, simulation, help);
+		
+		return Components.wrap(menuBar);
+	}
```
Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/Main.java]: This method is incredibly long. This was mentioned before, but we should refactor this in the future. (I have some ideas of how we can clean up this whole class - let's discuss them after this sprint)

File: src/edu/asu/plp/tool/prototype/Main.java [hunk 8]
```diff
@@ -413,45 +463,843 @@ private ProjectExplorerTree createProjectTree()
...
+	{
+		DropShadow ds = new DropShadow();
+		if (!simMode)
+		{
+			for (int x = 9; x <= 23; x++)
+			{
+				toolbar.getChildren().get(x).setEffect(null);
+				toolbar.getChildren().get(x).setDisable(false);
+			}
+			simMode = true;
```
Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/Main.java]: This is the same issue as line 633. The fact that this is repeated 3 times is a good reason to group the buttons in a set, then iterate over the set.

I presume these buttons live for the entire lifecycle of the JFrame; is that correct?

File: src/edu/asu/plp/tool/prototype/Main.java [hunk 9]
```diff
@@ -413,45 +463,843 @@ private ProjectExplorerTree createProjectTree()
...
+	{
+		Tab selectedTab = openProjectsPanel.getSelectionModel().getSelectedItem();
+		ASMFile activeFile = openProjects.getKey(selectedTab);
+		// TODO: check activeFile for null-value
+		return activeFile.getProject();
+	}
+	
+	private void onSimProjectClicked(MouseEvent event, HBox toolbar)
+	{
+		DropShadow ds = new DropShadow();
```
Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/Main.java]: Naming conventions: use "dropShadow" instead of ds

File: src/edu/asu/plp/tool/prototype/Main.java [hunk 10]
```diff
@@ -413,45 +463,843 @@ private ProjectExplorerTree createProjectTree()
...
+	}
+	
+	private void onlineManualWeb()
+	{
+		try
+		{
+			if (Desktop.isDesktopSupported())
+			{
+				Desktop.getDesktop().browse(new URI(
+						"https://code.google.com/p/progressive-learning-platform/wiki/UserManual"));
```
Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/Main.java]: This isn't bad, but it may be easier to debug/maintain if the inner argument is instantiated on a separate line, like so:
```
URI webAddress = new URI("https://code.google.com/p/progressive-learning-platform/wiki/UserManual");
Desktop.getDesktop().browse(webAddress);
```

File: src/edu/asu/plp/tool/prototype/Main.java [hunk 11]
```diff
@@ -413,45 +463,843 @@ private ProjectExplorerTree createProjectTree()
...
+			}
```
Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/Main.java]: There is no else block, in case Desktop is not supported.

In the case that it isn't supported, we should alert the user of the situation, and give them the link to enter manually, if desired.

File: src/edu/asu/plp/tool/prototype/Main.java [hunk 12]
```diff
@@ -413,45 +463,843 @@ private ProjectExplorerTree createProjectTree()
...
+		}
+		catch (Exception e)
+		{
+			e.printStackTrace();
```
Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/Main.java]: We should alert the user instead of printing the stack trace. Something like this (note that e was renamed to "exception"):
```
Dialogues.showAlertDialogue(exception, "Unable to open webpage");
```

Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/Main.java]: We should also log the exception

File: src/edu/asu/plp/tool/prototype/Main.java [hunk 13]
```diff
@@ -413,45 +463,843 @@ private ProjectExplorerTree createProjectTree()
...
+			@Override
+			public void handle(ActionEvent e)
+			{
+				String chosenLocation = "";
+				FileChooser fileChooser = new FileChooser();
+				// fileChooser.setTitle("Choose Project Location");
+				DirectoryChooser directoryChooser = new DirectoryChooser();
+				//directoryChooser.getExtensionFilters().add(new
+				// FileChooser.ExtensionFilter("PLP files (*.plp)", "*.plp"));
+				// directoryChooser.setInitialDirectory(projTextField.getText());
```
Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/Main.java]: Before this goes to master, commented code needs to be removed, unless it's part of a comment to clarify existing code

File: src/edu/asu/plp/tool/prototype/Main.java [hunk 14]
```diff
@@ -413,45 +463,843 @@ private ProjectExplorerTree createProjectTree()
...
+				directoryChooser.setTitle("Choose Project Location");
+				File file = directoryChooser.showDialog(null);
+				// If Cancel is chosen, throws a null pointer, needs to be fixed
```
Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/Main.java]: This is deserving of a FIXME comment:
```
// FIXME: If Cancel is chosen, throws a null pointer, needs to be fixed
```
or at the very least a TODO

File: src/edu/asu/plp/tool/prototype/Main.java [hunk 15]
```diff
@@ -413,45 +463,843 @@ private ProjectExplorerTree createProjectTree()
...
+				{
+					alert.setTitle("Invalid Project Name");
+					alert.setHeaderText(null);
+					alert.setContentText("You entered and invalid Project Name");
+					alert.showAndWait();
+					
+				}
+				else if (fileName.equals(""))
+				{
+					alert.setTitle("Invalid Project Name");
```
Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/Main.java]: Should be "Invalid File Name"

File: src/edu/asu/plp/tool/prototype/Main.java [hunk 16]
```diff
@@ -413,45 +463,843 @@ private ProjectExplorerTree createProjectTree()
...
+			public void handle(ActionEvent e)
+			{
+				String projectName;
+				String fileName;
+				String projectLocation;
+				Alert alert = new Alert(AlertType.INFORMATION);
+				projectName = projTextField.getText();
+				fileName = sourceFileField.getText();
+				projectLocation = projLocationField.getText();
+				if (projectName.equals(""))
```
Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/Main.java]: These checks should include a null-value check (just in case) and use trim() and isEmpty() instead of .equals(""). As so:
```
if (projectName == null || projectName.trim().isEmpty())
```

File: src/edu/asu/plp/tool/prototype/Main.java [hunk 17]
```diff
@@ -413,45 +463,843 @@ private ProjectExplorerTree createProjectTree()
...
+				{
+					alert.setTitle("Invalid Project Name");
+					alert.setHeaderText(null);
+					alert.setContentText("You entered and invalid Project Name");
+					alert.showAndWait();
+
```
Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/Main.java]: Use Dialogues.showInfoDialogue(title, message) instead.
This helps keep styling consistent throughout the application.

File: src/edu/asu/plp/tool/prototype/Main.java [hunk 18]
```diff
@@ -413,45 +463,843 @@ private ProjectExplorerTree createProjectTree()
...
+					alert.setContentText("You entered and invalid Project Name");
+					alert.showAndWait();
+					
+				}
+				else if (fileName.equals(""))
+				{
+					alert.setTitle("Invalid Project Name");
+					alert.setHeaderText(null);
+					alert.setContentText("You entered and invalid File Name");
+					alert.showAndWait();
```
Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/Main.java]: Use Dialogues.showInfoDialogue(title, message) instead.

File: src/edu/asu/plp/tool/prototype/Main.java [hunk 19]
```diff
@@ -413,45 +463,843 @@ private ProjectExplorerTree createProjectTree()
...
+				}
+				else if(projectLocation.equals(""))
+				{
+					alert.setTitle("Invalid Project Loaction");
+					alert.setHeaderText(null);
+					alert.setContentText("You entered and invalid Project Locaction");
+					alert.showAndWait();
+
```
Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/Main.java]: See above (for condition and alert message)

File: src/edu/asu/plp/tool/prototype/Main.java [hunk 20]
```diff
@@ -413,45 +463,843 @@ private ProjectExplorerTree createProjectTree()
...
+					PLPSourceFile sourceFile = new PLPSourceFile(project, fileName);
+					project.add(sourceFile);
+					project.save();
+					projects.add(project);
+					openFile(sourceFile);
+					Stage stage = (Stage) createProject.getScene().getWindow();
+					stage.close();
+				}
+			}
+		});
```
Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/Main.java]: We should find a way to move this block to it's own method, and set the listener as follows:
```
createProject.setOnAction(this::onCreateProjectClicked);
```
And make a method with the signature:
```
public void onCreateProjectClicked(ActionEvent event)
```

Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/Main.java]: Perhaps this window is substantial enough to warrant its own class...
We should discuss this when we talk about refactoring (this can be put off until next sprint)

File: src/edu/asu/plp/tool/prototype/model/PLPProject.java [hunk 21]
```diff
@@ -200,8 +207,8 @@ public void saveLegacy()
 	@Override
 	public void saveAs(String filePath)
 	{
-		// TODO: implement
-		throw new UnsupportedOperationException("Not Yet Implemented");
```
Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/model/PLPProject.java]: Why was this removed and replaced with 2 blank lines?
UnsupportedOperationException should stay (along with the TODO comment) until the method is implemented.

File: src/edu/asu/plp/tool/prototype/model/PLPProject.java [hunk 22]
```diff
@@ -145,8 +147,13 @@ public PLPProject(String name)
...
+		File file = new File(this.getPath());
+		file.mkdirs();
+		System.out.println(this.getPath());
+		System.out.println(this.getName());
+		//System.out.println(this.getFileCount());
+
+		//throw new UnsupportedOperationException("Not Yet Implemented");
```
Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/model/PLPProject.java]: See below. This code:
```
// TODO: implement
throw new UnsupportedOperationException("Not Yet Implemented");
```
should not be deleted until the method is fully implemented.
If (for some reason) the method was implemented, but is broken, then the TODO should be replaced with something along the lines of:
```
// FIXME: [describe what's wrong]
```

Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/model/PLPProject.java]: In this case the method isn't finished being implemented, so the TODO and UnsupportedOperationException should stay.

File: src/edu/asu/plp/tool/prototype/model/PLPProject.java [hunk 23]
```diff
@@ -7,6 +7,8 @@
 
 import javafx.beans.property.SimpleStringProperty;
 import javafx.beans.property.StringProperty;
+import javafx.stage.DirectoryChooser;
+import javafx.stage.FileChooser;
```
Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/model/PLPProject.java]: PLPProject is a model class. It shouldn't have any GUI components or interactions (originating from this class), including FileChooser and DirectoryChooser.

These imports (and their usage) should be removed.

File: src/edu/asu/plp/tool/prototype/Main.java [hunk 24]
```diff
@@ -66,9 +92,11 @@
 	public static final long REVISION = 1;
 	public static final int DEFAULT_WINDOW_WIDTH = 1280;
 	public static final int DEFAULT_WINDOW_HEIGHT = 720;
+	public static boolean simMode = false;
```
Comment (zcmoore) [inline src/edu/asu/plp/tool/prototype/Main.java]: This shouldn't be static.

File: src/edu/asu/plp/tool/prototype/Main.java (hunk 5 above)
Comment (EHawks) [inline src/edu/asu/plp/tool/prototype/Main.java]: Ya, I agree, will do so

File: src/edu/asu/plp/tool/prototype/Main.java (hunk 10 above)
Comment (EHawks) [inline src/edu/asu/plp/tool/prototype/Main.java]: Good Idea, I like that

File: src/edu/asu/plp/tool/prototype/Main.java (hunk 15 above)
Comment (EHawks) [inline src/edu/asu/plp/tool/prototype/Main.java]: Fixed

File: src/edu/asu/plp/tool/prototype/Main.java (hunk 17 above)
Comment (EHawks) [inline src/edu/asu/plp/tool/prototype/Main.java]: Will do
